eng/common/TestResources/New-TestResources.ps1

```diff
@@ -626,8 +631,37 @@ try {
 
         $postDeploymentScript = $templateFile.originalFilePath | Split-Path | Join-Path -ChildPath "$ResourceType-resources-post.ps1"
         if (Test-Path $postDeploymentScript) {
-            Log "Invoking post-deployment script '$postDeploymentScript'"
-            &$postDeploymentScript -ResourceGroupName $ResourceGroupName -DeploymentOutputs $deploymentOutputs @PSBoundParameters
+            if ($SelfContainedPostScript) {
+                $selfContainedScript = $templateFile.originalFilePath | Split-Path | Join-Path -ChildPath "self-contained_$ResourceType-resources-post.ps1"
+                Log "Creating invokable post-deployment script '$selfContainedScript' from '$postDeploymentScript'"
+
+                $deserialized = @{}
+                foreach ($parameter in $PSBoundParameters.GetEnumerator()) {
+                    if ($parameter.Value -is [System.Management.Automation.SwitchParameter]) {
+                        $deserialized[$parameter.Key] = $parameter.Value.ToBool()
+                    } else {
+                        $deserialized[$parameter.Key] = $parameter.Value
+                    }
+                }
+                $deserialized['ResourceGroupName'] = $ResourceGroupName
+                $deserialized['DeploymentOutputs'] = $deploymentOutputs
+                $serialized = $deserialized | ConvertTo-Json
+
+                $outScript = @"
+`$parameters = `@'
+$serialized
+'`@ | ConvertFrom-Json -AsHashtable
```

Comment:
Why serialize in the script as opposed to serializing the entire command line into a string to write to the script?

Reply:
The only way I can think of to convert instances arrays/objects and switches into powershell literals is via json serialization.

Reply:
Or rather, the json conversion avoids having to de-compile runtime powershell into a valid script.